[Draft] Remove `node-fetch`

brianjenkins94 opened the pull request.

mistval commented:
This looks well on its way. I realized the instructions in the README for running tests was out of date, so I just pushed a change for that, but it sounds like you might have already figured out how to run them.

brianjenkins94 commented:
Actually, I think the "Has a url property" and "Has a redirected property" tests might be `node-fetch`-specific.

So I think this is ready for review.

A couple of things that I removed:
- `.buffer` as a method on the `Response` (legacy, use `.arrayBuffer` instead)
- `NodeJS.Buffer` usage anywhere (prefer browser global `window.Buffer`)
- `fs.ReadStream` support (this could be put back, I was just already wrestling with `NodeJS.ReadableStream` vs. Web Streams API `ReadableStream` types and that was about all my brain could handle; the related tests were commented-out)
- I switched the md5 hash to a sha1 hash that uses the browser global `window.crypto` (I would have thought Web Crypto had md5 but it doesn't)

The motivation for these changes was to gently push towards web standards and maybe a version that could work in the browser (`cacache` may make that difficult/impossible). Interop with the [cacheable](https://github.com/jaredwray/cacheable) ecosystem would be cool though.

The ~~`url`, `redirected` and~~ `size` property may no longer be populated on the `Response` object after removing `node-fetch`.

I was wrong, [`url`](https://developer.mozilla.org/en-US/docs/Web/API/Response/url) and [`redirected`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirected) should be there. The `size` property is definitely gone though.

mistval commented on src/classes/response.ts:
```diff
@@ -1,24 +1,19 @@
-import { Readable } from 'stream';
-import type { Response as NodeFetchResponseType, ResponseInit as NodeFetchResponseInit } from 'node-fetch';
 import { NFCResponseMetadata } from '../types.js';
-import { getNodeFetch } from '../helpers/node_fetch_imports.js';
```
Could you delete the `node_fetch_imports.ts` file?

brianjenkins94 replied:
Done.

mistval commented on plugins/redis/redis_cache.ts:
```diff
@@ -46,7 +45,7 @@ export class RedisCache implements INodeFetchCacheCache {
 
     return {
-      bodyStream: readableStream,
+      bodyStream: readableStream as Omit<ReadableStream<any>, "closed">,
```
Would `<unknown>` work here instead of `<any>`? I'd prefer using `unknown` over `any` where possible.

brianjenkins94 replied:
I was able to get rid of this.

mistval commented on plugins/redis/redis_cache.ts:
```diff
@@ -57,7 +56,7 @@ export class RedisCache implements INodeFetchCacheCache {
   }
 
-  async set(key: string, bodyStream: NodeJS.ReadableStream, metaData: NFCResponseMetadata) {
+  async set(key: string, bodyStream: Omit<ReadableStream, "closed">, metaData: NFCResponseMetadata) {
```
Why do we `Omit` the `closed` property?

brianjenkins94 replied:
I was able to get rid of this.

mistval commented on plugins/redis/test_redis_cache.ts:
```diff
@@ -329,7 +329,8 @@ describe('REDIS Plugin Tests', function() {
       response = await defaultCachedFetch(TWO_HUNDRED_URL, post(s1));
       assert.strictEqual(response.returnedFromCache, true);
     });
```
Using fs.ReadStream already has some caveats in the current version and is probably a minor use case so it's not critical, but I do think we should make an effort to keep that support. I can try to take a look if you need a second pair of eyes on how to restore that.

brianjenkins94 replied:
So right now, without changing the tests, attempting to use a NodeJS Stream throws:

```
TypeError: RequestInit: duplex option is required when sending a body.
```

As I understand it, we need to add `duplex: "half"` when there is a NodeJS Stream body. Should this be a breaking-change left for the user? (more spec-compliant) Or should we quietly add `duplex: body !== undefined ? "half" : undefined` or something for maximum compatibility?

Or I can just change the tests.

mistval replied:
You would need to add `duplex: 'half'` even if you were using native fetch directly and not using this package, right? If that's the case I'd say we should require the user to specify it, and we'll note it in the upgrade guide.

brianjenkins94 replied:
Correct. I changed it in the test files.

mistval commented on plugins/redis/test_redis_cache.ts:
```diff
@@ -367,7 +368,7 @@ describe('REDIS Plugin Tests', function() {
     it('Can use a client-provided custom cache key', async () => {
       const cacheFunction = async (resource: FetchResource) => {
-        if (resource instanceof StandardFetchRequest) {
+        if (resource instanceof Request) {
```
Could you rename this back to `StandardFetchRequest`? With an assignment at the top of the file like `const StandardFetchRequest = Request;`

I might link to test diff in upgrade instructions so it would be ideal if the diff were minimal.

brianjenkins94 replied:
Across the whole project? or just this file?

mistval replied:
Just in the test case files.

brianjenkins94 replied:
Done.

mistval commented on src/classes/response.ts:
```diff
@@ -1,24 +1,19 @@
-        headers: response.headers.raw(),
-        size: response.size,
+        headers: Array.from(response.headers.entries()).reduce<Record<string, string[]>>((headers, [key, value]) => { headers[key] = [...(headers[key] ?? []), value]; return headers; }, {}),
+        //size: response.size,
```
Seems like this doesn't exist on native fetch Response so let's remove this line entirely instead of commenting it.

brianjenkins94 replied:
Done.

mistval commented on src/classes/response.ts:
```diff
@@ -1,24 +1,19 @@
         statusText: response.statusText,
-        headers: response.headers.raw(),
-        size: response.size,
+        headers: Array.from(response.headers.entries()).reduce<Record<string, string[]>>((headers, [key, value]) => { headers[key] = [...(headers[key] ?? []), value]; return headers; }, {}),
```
This line's a bit long and hard to read, it would benefit from being spread across a few lines.

brianjenkins94 replied:
Somewhat improved. I think this and consuming streams happens a couple of places in the codebase, so it would probably benefit from being its own function in its own file.

mistval commented on src/index.ts:
```diff
@@ -43,15 +40,20 @@ async function getResponse(
 
+  if (originalResource instanceof Request) {
+    resource = originalResource.clone()
+  }
```
How come we need to clone it here? You probably noticed already but this cloning is probably why it wasn't working with fs.ReadStream (or at least is one of the reasons).

brianjenkins94 replied:
Yeah, this was weird to me too. But without it `fetch`/`undici` throws:

```
TypeError: Cannot construct a Request with a Request object that has already been used.
```

mistval replied:
Oh interesting, perhaps `undici` starts consuming the stream as soon as the `Request` is instantiated, before it gets sent via `fetch()`.

brianjenkins94 replied:
Yeah, this could use a little more investigation to see what specifically is causing the problem and whether it is a `undici` bug, but I'm reasonably happy with it the way it is.

mistval commented:
Thanks for your great work on this. I'd like to merge it after we review. It would be a major version increment so we would need an upgrade guide also but I could write that. It probably wouldn't be comprehensive since it will be impossible to know every difference between node-fetch and native fetch, but the changes to tests will be a good foundation for knowing the major differences.

Native fetch Response does seem to have `url` and `redirected` properties, I can see them printed when I run `await fetch('http://google.com')` in the node v22 REPL. We may just need to assign those properties to `this` explicitly in the `NFCResponse` constructor.

brianjenkins94 commented:
I'm not sure where to make the change for the `url` and `redirected` properties.

mistval commented:
> I'm not sure where to make the change for the `url` and `redirected` properties.

I pushed a change to your branch for it. The properties weren't writable like I was hoping (they were getters) so I used `Object.defineProperty()` and overrode `clone()` so that they would get copied over if you clone the response (might be nice to have `clone()` return an `NFCResponse` actually, but that's probably one for another day).

All tests are passing now! I will try to take a closer look Sunday or Monday and do an upgrade guide. I might have some more comments but I think this is likely at least 95% of the way there.

brianjenkins94 commented:
I was trying something with:

```
public override url;

constructor(/* ... */) {
  super(/* ... */)
  this.url = metaData.url;
}
```

that had worked for `url`, but hadn't quite worked for `redirected`, which might be marginally cleaner.

> All tests are passing now! I will try to take a closer look Sunday or Monday and do an upgrade guide.

Thanks! This was fun (definitely more challenging than I had originally thought)!

This will be my first real PR to an open-source project 🎉

brianjenkins94 commented:
You may also be able to replace `formdata-node` with the [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) browser global.

brianjenkins94 commented:
I can also commit my `tsup.config.json` to help with building. Handling `import("stream/web")` is quite a pain.

mistval commented on src/classes/response.ts:
```diff
@@ -57,9 +57,16 @@ async function createNFCResponseClass() {
+    }
+
+    public override clone(): Response {
+      return new NFCResponse(super.body as ReadableStream, this.metaData, this.ejectFromCache, this.returnedFromCache, this.isCacheMiss)
```
This won't quite work because it doesn't clone the body stream, so if you clone a response and then try to read the bodies of both the original and the clone, it will error. I just pushed up a test for that, currently failing.

However the more I think about it, I do think it's worthwhile to try to find a solution that returns an NFCResponse here from `clone()`. This approach appears to work:

```ts
    _url = '';
    _redirected: boolean = false

    override get url() {
      return this._url;
    }

    override get redirected() {
      return this._redirected;
    }

    constructor(
      bodyStream: ReadableStream,
      public metaData: Omit<ResponseInit, 'headers'> & {
        url: string;
        redirected: boolean;
        counter: number;
        headers: Record<string, string[]>;
      },
      public ejectFromCache: () => Promise<unknown>,
      public returnedFromCache: boolean,
      public isCacheMiss = false,
    ) {
      super(
        bodyStream,
        metaData as any
      );

      this._url = metaData.url;
      this._redirected = metaData.redirected
    }

    public override clone(): NFCResponse {
      const superClone = super.clone() as NFCResponse;

      Object.setPrototypeOf(superClone, NFCResponse.prototype);

      superClone._url = this.url;
      superClone._redirected = this.redirected;
      superClone.ejectFromCache = this.ejectFromCache;
      superClone.returnedFromCache = this.returnedFromCache;
      superClone.isCacheMiss = this.isCacheMiss;

      return superClone;
    }
```

It would also be possible to replace the `createNFCResponseClass()` function with just a normal class export. That hackery was only needed to support CommonJS using dynamic imports from node-fetch.

mistval replied:
This may be preferable:

```ts
public override clone(): NFCResponse {
      const clonedBody = super.clone();
    
      return new NFCResponse(
        clonedBody.body as ReadableStream,
        {
          url: this.url,
          redirected: this.redirected,
          status: this.status,
          statusText: this.statusText,
          headers: NFCResponse.serializeHeaders(this),
          counter: this.metaData.counter,
        },
        this.ejectFromCache,
        this.returnedFromCache,
        this.isCacheMiss,
      );
    }
```

Due to the [caveats about changing an object's prototype](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf).

brianjenkins94 replied:
Definitely reads better than `Object.defineProperty()`.

mistval commented:
> I can also commit my `tsup.config.json` to help with building. Handling `import("stream/web")` is quite a pain.

When would this be needed? It seems to be building okay without that, when I run `npm run tsc && npm run buildcjs`.

brianjenkins94 commented:
I think in the output you'll find `import { ReadableStream } from "stream/web";` which would fail if run in the browser.

Although, now that I think about it, you can just replace all instances of `ReadableStream` with `globalThis.ReadableStream` and continue to use your existing build process.

mistval commented:
> I think in the output you'll find `import { ReadableStream } from "stream/web";` which would fail if run in the browser.

Ah maybe this module can support browser someday, but for now it's not necessary.

LeaVerou commented:
Hi @mistval, this PR seems to have stalled, but from reading the discussion it seems like it was fairly mature. Are there any blockers?

mistval commented:
> Hi @mistval, this PR seems to have stalled, but from reading the discussion it seems like it was fairly mature. Are there any blockers?

Hi @LeaVerou , I do recall this was really close. At a minimum, whatever tests were failing need to be fixed, and README needs some updates. I'll try to take a closer look soon, I'll need to refresh myself on it.
